flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/TableFunctionCollector.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.collector;
+
+import org.apache.flink.api.common.functions.AbstractRichFunction;
+import org.apache.flink.table.functions.TableFunction;
+import org.apache.flink.util.Collector;
+
+/**
+ * The basic implementation of collector for {@link TableFunction}.
+ */
+public abstract class TableFunctionCollector<T> extends AbstractRichFunction implements Collector<T> {
+
+	private static final long serialVersionUID = 1L;
+
+	private Object input;
+	private Collector<?> collector;
+	private boolean collected;
+
+	/**
+	 * Sets the input row from left table,
+	 * which will be used to cross join with the result of table function.
+	 */
+	public void setInput(Object input) {
+		this.input = input;
+	}
+
+	/**
+	 * Gets the input value from left table,
+	 * which will be used to cross join with the result of table function.
+	 */
+	public Object getInput() {
+		return input;
+	}
+
+	/**
+	 * Sets the current collector, which used to emit the final row.
+	 */
+	public void setCollector(Collector<?> collector) {
+		this.collector = collector;
+	}
+
+	/**
+	 * Gets the internal collector which used to emit the final row.
+	 */
+	public Collector<?> getCollector() {
+		return collector;
+	}
+
+	/**
+	 * Resets the flag to indicate whether [[collect(T)]] has been called.
+	 */
+	public void reset() {
+		this.collected = false;
+	}
+
+	/**
+	 * Whether {@link #collect(Object)} has been called.
+	 *
+	 * @return True if {@link #collect(Object)} has been called.
+	 */
+	public boolean isCollected() {
+		return collected;
+	}
+
+	@Override
+	public void collect(T record) {
+		this.collected = true;
```

Comment:
Shouldn't this method call `collector.collect`?

Reply:
and i just found that the sub-class of this class `RowToBaseRowCollector` is actually calling `      getCollector.asInstanceOf[Collector[BaseRow]].collect(result)`,

Reply:
What i meant is this class already holds a `private Collector<?> collector;`, and also implements `Collector<T>` interface. In this `collect(T record)` method, why don't you collect the record with your own collector, but rely on the caller to first `getCollector()` and collect records?

Reply:
Because `TableFunctionCollector` abstract class does't know the what is the record type `T`. It might be a `BaseRow`, but can also be a `Row` (see `RowToBaseRowCollector`). As a result, `TableFunctionCollector` can't collect by himself, because he doesn't know what the result record is.

Reply:
Then i think this class shouldn't be a collector, aka shouldn't implementing `Collector<T>` interface

Reply:
We have to keep the `collect(T record)` interface, because it used to accept UDTF's result value.

What about provide a `outputResult` method used to collect final result. And remove `getCollector()` method and `collect(T record)` implementation. 

```java
	public void outputResult(Object result) {
		this.collected = true;
		collector.collect(result);
	}
```

All the sub-classes can use `outputResult(result)` to collect final result instead of `getCollector().collect(result)`.

Reply:
yes, let's try this.